src/System.Memory/src/System/ReadOnlySpan.Portable.cs

```diff
@@ -193,6 +193,10 @@ public override string ToString()
         {
             if (typeof(T) == typeof(char))
             {
+                if (_pinnable is string str)
```

Comment:
Note that the optimizing compilers can figure out here that `Pinnable<T>` can never be a `string`, and turn this condition into false.

I think you may want to pass it through `Unsafe.As` here to strip the type information here. It is not guaranteed to work, but it is increasing the changes that it will.

Reply:
We may want to have a slow-span specific test for this so that we will know when it does not work.

Reply:
> Note that the optimizing compilers can figure out here that Pinnable<T> can never be a string

Even the C# compiler will... that's why in my example I stored it into an object first, otherwise the C# compiler just emit the cast as a hard-coded false. Though as you say an even better compiler could see through the object cast as well.